OsmAnd/src/net/osmand/plus/routing/RouteCalculationResult.java

```diff
@@ -284,9 +287,32 @@ private static void attachAlarmInfo(List<AlarmInfo> alarms, RouteSegmentResult r
 				loc.setLatitude(MapUtils.get31LatitudeY(y31));
 				loc.setLongitude(MapUtils.get31LongitudeX(x31));
 				AlarmInfo info = AlarmInfo.createAlarmInfo(typeRule, locInd, loc);
+
 				// For STOP first check if it has directional info
 				if ((info != null) && !((info.getType() == AlarmInfoType.STOP) && !res.getObject().isStopApplicable(res.isForwardDirection(), intId, res.getStartPointIndex(), res.getEndPointIndex()))) {
 					alarms.add(info);
+
+					if (info.getType() == AlarmInfoType.SPEED_CAMERA) {
+						speedCameraAlarmInfo = info;
+					}
+				}
+
+				if (maxSpeed <= 0 || maxSpeed == RouteDataObject.NONE_MAX_SPEED) {
+					maxSpeed = typeRule.maxSpeed();
+				}
+			}
+
+			if (speedCameraAlarmInfo != null) {
+				if (maxSpeed <= 0 || maxSpeed == RouteDataObject.NONE_MAX_SPEED) {
+					maxSpeed = res.getObject().getMaximumSpeed(res.isForwardDirection());
+				}
+
+				if (maxSpeed > 0 && maxSpeed != RouteDataObject.NONE_MAX_SPEED) {
+					if (sc.imperial) {
```

Comment:
Looks strange that here we store converted data, I guess it should happen at the end UI / Voice prompt cause here it's to early to do conversions

Reply:
@vshcherb To be honest, I also had strong doubts when writing this code.

On the right hand it is not the best place for conversion, on the left hand the `AlarmInfo.intValue` also used for speed limit alarm, and in that case the field will contain converted value.

It could be very confusing if the same field will contain speed value in different units of measurement.

Probably, it would be better to introduce dedicated explicit field in AlarmInfo for meters per second speed and provide getter method to provide converted speed. But it will require re-factoring in multiple places, and also I guess there was signifact reason to use several common fields like "floatValue", "intValue" etc. instead of "maxSpeedInMetersPerSecond", "distance" etc...